From a9ec49b4e728009c4bb4a2af7cb80a3c1ce6016c Mon Sep 17 00:00:00 2001
From: Cameron Gutman <aicommander@gmail.com>
Date: Thu, 12 Jan 2023 17:49:29 -0600
Subject: [PATCH] lavu/hwcontext_qsv: fix MFX session leak in hwdevice

There was no device_uninit() implementation to handle freeing
the resources allocated in qsv_device_derive_from_child().
---
 libavutil/hwcontext_qsv.c | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/libavutil/hwcontext_qsv.c b/libavutil/hwcontext_qsv.c
index 42851d4fd5..ad2e9ffc9d 100644
--- a/libavutil/hwcontext_qsv.c
+++ b/libavutil/hwcontext_qsv.c
@@ -266,6 +266,17 @@ static int qsv_fill_border(AVFrame *dst, const AVFrame *src)
     return 0;
 }
 
+static void qsv_device_uninit(AVHWDeviceContext *ctx)
+{
+    AVQSVDeviceContext *hwctx = ctx->hwctx;
+
+    if (hwctx->session)
+        MFXClose(hwctx->session);
+
+    if (hwctx->loader)
+        MFXUnload(hwctx->loader);
+}
+
 static int qsv_device_init(AVHWDeviceContext *ctx)
 {
     AVQSVDeviceContext *hwctx = ctx->hwctx;
@@ -2217,6 +2228,7 @@ const HWContextType ff_hwcontext_type_qsv = {
     .device_create          = qsv_device_create,
     .device_derive          = qsv_device_derive,
     .device_init            = qsv_device_init,
+    .device_uninit          = qsv_device_uninit,
     .frames_get_constraints = qsv_frames_get_constraints,
     .frames_init            = qsv_frames_init,
     .frames_uninit          = qsv_frames_uninit,
